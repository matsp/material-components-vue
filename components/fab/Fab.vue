<template>
  <button
    v-if="href.length === 0"
    :class="classes"
    class="mdc-fab"
    v-bind="$attrs"
    v-on="$listeners"
  >
    <slot />
  </button>
  <a
    v-else
    :href="href"
    :class="classes"
    class="mdc-fab"
    v-bind="$attrs"
    v-on="$listeners"
  >
    <slot />
  </a>
</template>

<script>
import { MDCRipple } from '@material/ripple'

import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  props: {
    mini: {
      type: Boolean,
      default: false
    },
    absoluteRight: {
      type: Boolean,
      default: false
    },
    exited: {
      type: Boolean,
      default: false
    },
    ripple: {
      type: Boolean,
      default: true
    },
    href: {
      type: String,
      default: ''
    }
  },
  data () {
    return {
      mdcRipple: undefined,
      extended: false,
      slotObserver: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-fab--mini': this.mini,
        'mdc-fab--absolute-right': this.absoluteRight,
        'mdc-fab--exited': this.exited,
        'mdc-fab--extended': this.extended
      }
    }
  },
  watch: {
    classes () {
      this.reInstantiateRipple()
    },
    ripple () {
      this.reInstantiateRipple()
    }
  },
  mounted () {
    this.updateSlot()
    this.slotObserver = new MutationObserver(() => this.updateSlot())
    this.slotObserver.observe(this.$el, {
      childList: true,
      subtree: true
    })
    if (this.ripple) this.mdcRipple = MDCRipple.attachTo(this.$el)
  },
  beforeDestroy () {
    if (this.mdcRipple) {
      this.mdcRipple.destroy()
    }
    this.slotObserver.disconnect()
  },
  methods: {
    updateSlot () {
      if (this.$slots.default) {
        const defaults = this.$slots.default
        let i = 0
        for (; i < defaults.length; i++) {
          if (defaults[i].elm instanceof Element && defaults[i].elm.classList.contains('mdc-fab__label')) {
            this.extended = true
            break
          }
        }
        if (i === defaults.length) this.extended = false // no label found
      }
    },
    reInstantiateRipple () {
      if (this.ripple) {
        if (this.mdcRipple) {
          this.mdcRipple.destroy()
        }
        MDCRipple.attachTo(this.$el)
      } else {
        if (this.mdcRipple) {
          this.mdcRipple.destroy()
        }
        this.mdcRipple = undefined
      }
    }
  }
}
</script>
