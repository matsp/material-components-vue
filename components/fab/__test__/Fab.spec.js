import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import Fab from '../Fab.vue'
import Vue from 'vue'

describe('Fab', () => {
  it('should mount', () => {
    const wrapper = mount(Fab)
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.$data.mdcRipple).toBeDefined()
    expect(wrapper.vm.$data.slotObserver).toBeDefined()
  })

  it('should render with no prop', () => {
    const wrapper = mount(Fab)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-fab')
  })

  it('should render as mini', () => {
    const wrapper = mount(Fab, {
      propsData: {
        mini: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-fab--mini')
  })

  it('should render as exited', () => {
    const wrapper = mount(Fab, {
      propsData: {
        exited: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-fab--exited')
  })

  it('should render as extended', async () => {
    const wrapper = mount(Fab, {
      slots: {
        default: 'create'
      }
    })
    await Vue.nextTick()
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-fab--extended')
  })

  it('should render as absolute right', () => {
    const wrapper = mount(Fab, {
      propsData: {
        absoluteRight: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-fab--absolute-right')
  })

  it('should render without ripple', () => {
    const wrapper = mount(Fab, {
      propsData: {
        ripple: false
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.vm.$data.mdcRipple).toBeUndefined()
  })
})
