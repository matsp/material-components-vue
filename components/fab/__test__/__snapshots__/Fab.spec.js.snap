// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fab Icon should render with no prop 1`] = `<span class="mdc-fab__icon">favorite</span>`;

exports[`Fab Label should render with no prop 1`] = `<span class="mdc-fab__label"></span>`;

exports[`Fab should render as absolute right 1`] = `<button class="mdc-fab mdc-fab--absolute-right"></button>`;

exports[`Fab should render as exited 1`] = `<button class="mdc-fab mdc-fab--exited"></button>`;

exports[`Fab should render as extended 1`] = `<button class="mdc-fab mdc-fab--extended"><span class="mdc-fab__label"></span></button>`;

exports[`Fab should render as mini 1`] = `<button class="mdc-fab mdc-fab--mini"></button>`;

exports[`Fab should render with no prop 1`] = `<button class="mdc-fab"></button>`;

exports[`Fab should render without ripple 1`] = `<button class="mdc-fab"></button>`;
